### PR TITLE
i#4301: Fix x86-32 Linux stack alignment before dynamo_process_exit

### LINKED ISSUE #4301: Call dynamorio_process_exit is not 16 byte aligned (Linux x86-32)
Hi!

I run my x86-32 DynamoRIO client on Linux (Release build). I get unaligned stack here:
```
0xf7eb81f3 <cat_done_saving_dstack+16>  call   0xf7ddfaf0 <dynamo_process_exit>

#13 0xf7e3fc26 in instrument_exit_event () at /data/predicate_solving/dynamorio/core/lib/instrument.c:871
#14 0xf7dcbdd0 in dynamo_process_exit () at /data/predicate_solving/dynamorio/core/dynamo.c:1554
#15 0xf7ea41f8 in cat_done_saving_dstack ()
#16 0xf7ea4257 in cat_have_lock ()
```

This leads to further crash in my client due to XMM usage.
```
0xf663d332 <_ZN15prime_generator22process_next_k_numbersEy+290> movaps XMMWORD PTR [ebp-0xe8],xmm3
```

### PR DESCRIPTION
**Pull request by SweetVishnya**
Closes #4301

### CODE REVIEW

File: core/arch/x86/x86.asm [hunk 1]
```diff
@@ -604,6 +604,9 @@ cat_done_saving_dstack:
         CALLC0(GLOBAL_REF(get_cleanup_and_terminate_global_do_syscall_entry))
 #if defined(MACOS) && !defined(X64)
         lea      REG_XSP, [-2*ARG_SZ + REG_XSP] /* maintain align-16 w/ 2 pushes below */
+#endif
+#if defined(LINUX) && !defined(X64)
+        lea      REG_XSP, [-ARG_SZ + REG_XSP] /* maintain align-16 w/ 2 pushes below */
```
Comment (derekbruening) [inline core/arch/x86/x86.asm:609]: If you could help me understand: why is Linux different from Mac?  At a glance aren't they the same below with 2 pushes?

File: core/arch/x86/x86.asm [hunk 2]
```diff
@@ -649,6 +652,9 @@ cat_have_lock:
         mov      REG_XSI, [2*ARG_SZ + REG_XBP]  /* sysnum */
         pop      REG_XAX             /* syscall */
         pop      REG_XCX             /* dstack */
+#if defined(LINUX) && !defined(X64)
+        lea      REG_XSP, [ARG_SZ + REG_XSP] /* undo align-16 lea from above */
+#endif
```
Comment (derekbruening) [inline core/arch/x86/x86.asm:657]: I wonder if you could add an exit event that calls check_stack_alignment() to suite/tests/client-interface/cleancall.dll.c and confirm it crashes without this fix? Then we'll have a regression test.

Comment (SweetVishnya) [reply core/arch/x86/x86.asm:657]: How can I run this specific test?

File: core/arch/x86/x86.asm (hunk 1 above)
Comment (SweetVishnya) [reply core/arch/x86/x86.asm:609]: It seems to me that stack is misaligned for previous call too. For instance, `get_cleanup_and_terminate_global_do_syscall_entry`. I need further investigation.

Comment (SweetVishnya) [reply core/arch/x86/x86.asm:609]: I will continue on this pull request after some sleep.

File: core/arch/x86/x86.asm (hunk 2 above)
Comment (derekbruening) [reply core/arch/x86/x86.asm:657]: `ctest -V -R 'client\.cleancall$'`

File: core/arch/x86/x86.asm (hunk 1 above)
Comment (derekbruening) [reply core/arch/x86/x86.asm:609]: Sure, np

**Comment by SweetVishnya** [comment]
Done.

File: core/arch/x86/x86.asm [hunk 3]
```diff
@@ -567,7 +567,7 @@ GLOBAL_LABEL(cleanup_and_terminate:)
         mov      [4*ARG_SZ + REG_XBP], ARG4
 #else
         mov      REG_XBP, REG_XSP
-# if defined(MACOS) && !defined(X64)
+# ifdef UNIX
```
Comment (derekbruening) [inline core/arch/x86/x86.asm:570]: So this changes the 64-bit behavior: is that intentional?  You found that the 64-bit alignment was off too?

File: core/arch/x86/x86.asm [hunk 4]
```diff
@@ -1791,11 +1791,11 @@ GLOBAL_LABEL(dr_setjmp:)
         /* PR 206278: for try/except we need to save the signal mask */
         mov      REG_XDX, ARG1
         push     REG_XDX /* preserve */
-# if defined(MACOS) && !defined(X64)
+# ifndef X64
```
Comment (derekbruening) [inline core/arch/x86/x86.asm:1794]: This and the #if below used to be Mac-only and unlike all the other changes in this PR here it's expanded to Windows too: is that intentional?  32-bit Windows has no stack alignment beyond 4 in its ABI.  I expected all the changes to be s/MACOS/UNIX/?

File: core/arch/x86/x86.asm (hunk 3 above)
Comment (SweetVishnya) [reply core/arch/x86/x86.asm:570]: This affects only 32-bit, it is the `#else` of `#ifdef X64`.

File: core/arch/x86/x86.asm (hunk 4 above)
Comment (SweetVishnya) [reply core/arch/x86/x86.asm:1794]: this affects only UNIX, see `#ifdef UNIX` above

**Comment by SweetVishnya** [comment]
So, mainly I just exploit nested ifdefs. No need to recheck UNIX. Also, I fixed Windows CI.

**Comment by AssadHashmi** [comment]
run arm tests

File: core/arch/x86/x86.asm (hunk 3 above)
Comment (derekbruening) [reply core/arch/x86/x86.asm:570]: OK, you can see how from the PR diff view it looks suspect.

File: core/arch/x86/x86.asm (hunk 4 above)
Comment (derekbruening) [reply core/arch/x86/x86.asm:1794]: OK.

**Comment by SweetVishnya** [comment]
OK, I will do so in future.

Fix for Windows CI was adding explicit (void) in function arguments.

**Pull request merged**